File: plugin/tongs-runner/src/main/java/com/github/tarcv/tongs/TongsCli.java
/*
 * Copyright 2019 TarCV
 * Copyright 2014 Shazam Entertainment Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License.
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.github.tarcv.tongs;

import com.beust.jcommander.*;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;

import static com.github.tarcv.tongs.Configuration.Builder.configuration;
import static com.github.tarcv.tongs.injector.GsonInjector.gson;
import static com.github.tarcv.tongs.utils.Utils.cleanFile;

public class TongsCli {

    private static final Logger logger = LoggerFactory.getLogger(TongsCli.class);

    private TongsCli() {}

    public static class CommandLineArgs {

        @Parameter(names = { "--sdk" }, description = "Path to Android SDK", converter = FileConverter.class)
        public File sdk;

        @Parameter(names = { "--apk" }, description = "Application APK", converter = FileConverter.class,
                required = true)
        public File apk;

        @Parameter(names = { "--test-apk" }, description = "Test application APK", converter = FileConverter.class,
                required = true)
        public File testApk;

        @Parameter(names = { "--config" }, description = "Path of JSON config file", converter = FileConverter.class,
                required = true)
        public File configurationFile;

        @Parameter(names = { "-h", "--help" }, description = "Command help", help = true, hidden = true)
        public boolean help;
    }

    /* JCommander deems it necessary that this class be public. Lame. */
    public static class FileConverter implements IStringConverter<File> {

        @Override
        public File convert(String s) {
            return cleanFile(s);
        }
    }

    public static void main(String... args) {
        CommandLineArgs parsedArgs = new CommandLineArgs();
        JCommander jc = new JCommander(parsedArgs);

        try {
            jc.parse(args);
        } catch (ParameterException e) {
            StringBuilder out = new StringBuilder(e.getLocalizedMessage()).append("\n\n");
            jc.usage(out);
            logger.error(out.toString());
            System.exit(1);
            return;
        }
        if (parsedArgs.help) {
            jc.usage();
            return;
        }

        try {
            Reader configFileReader = new FileReader(parsedArgs.configurationFile);
            TongsConfigurationJsonExtension tongsConfiguration = gson().fromJson(configFileReader, TongsConfigurationJsonExtension.class);

            Configuration configuration = configuration()
                    .withAndroidSdk(parsedArgs.sdk != null ? parsedArgs.sdk : cleanFile(CommonDefaults.ANDROID_SDK))
                    .withApplicationApk(parsedArgs.apk)
                    .withInstrumentationApk(parsedArgs.testApk)
                    .withOutput(tongsConfiguration.baseOutputDir != null ? cleanFile(tongsConfiguration.baseOutputDir) : cleanFile(Defaults.TONGS_OUTPUT))
                    .withTitle(tongsConfiguration.title)
                    .withSubtitle(tongsConfiguration.subtitle)
                    .withTestClassRegex(tongsConfiguration.testClassRegex)
                    .withTestPackage(tongsConfiguration.testPackage)
                    .withTestOutputTimeout(tongsConfiguration.testOutputTimeout)
                    .withTestSize(tongsConfiguration.testSize)
                    .withExcludedSerials(tongsConfiguration.excludedSerials)
                    .withFallbackToScreenshots(tongsConfiguration.fallbackToScreenshots)
                    .withTotalAllowedRetryQuota(tongsConfiguration.totalAllowedRetryQuota)
                    .withRetryPerTestCaseQuota(tongsConfiguration.retryPerTestCaseQuota)
                    .withCoverageEnabled(tongsConfiguration.isCoverageEnabled)
                    .withPoolingStrategy(tongsConfiguration.poolingStrategy)
                    .withAutoGrantPermissions(tongsConfiguration.autoGrantPermissions)
                    .withExcludedAnnotation(tongsConfiguration.excludedAnnotation)
                    .build();

            Tongs tongs = new Tongs(configuration);
            if (!tongs.run() && !tongsConfiguration.ignoreFailures) {
                System.exit(1);
            }
        } catch (FileNotFoundException e) {
            logger.error("Could not find configuration file", e);
            System.exit(1);
        }
    }
}

File: plugin/cli/src/main/java/com/github/tarcv/tongs/cli/TongsCli.kt
/*
 * Copyright 2022 TarCV
 * Copyright 2014 Shazam Entertainment Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License.
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.github.tarcv.tongs.cli

import com.beust.jcommander.IStringConverter
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import com.beust.jcommander.ParameterException
import com.github.tarcv.tongs.CommonDefaults
import com.github.tarcv.tongs.Configuration
import com.github.tarcv.tongs.Tongs
import com.github.tarcv.tongs.TongsConfigurationJsonExtension
import com.github.tarcv.tongs.Utils
import com.github.tarcv.tongs.injector.GsonInjector
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileNotFoundException
import kotlin.system.exitProcess

object TongsCli {
    private val logger = LoggerFactory.getLogger(TongsCli::class.java)

    @JvmStatic
    fun main(args: Array<String>) {
        val parsedArgs = CommandLineArgs()
        JCommander(parsedArgs).let { jc ->
            try {
                jc.parse(*args)
            } catch (e: ParameterException) {
                StringBuilder(e.localizedMessage)
                    .append("\n\n").let {
                        jc.usage(it)
                        logger.error(it.toString())
                    }
                exitProcess(1)
            }
            if (parsedArgs.help) {
                StringBuilder().let {
                    jc.usage(it)
                    logger.error(it.toString())
                }
                exitProcess(0)
            }
        }

        try {
            val tongsConfiguration = parsedArgs.configurationFile
                    .let { 
                        if (it == "-") {
                            System.`in`.bufferedReader(Charsets.UTF_8)
                        } else {
                            requireNotNull(FileConverter().convert(it)) {"Configuration should be set" }
                                .bufferedReader(Charsets.UTF_8)
                        }
                    }
                    .use { configFileReader ->
                        GsonInjector.gson().fromJson(configFileReader, TongsConfigurationJsonExtension::class.java)
                    }

            val configuration = Configuration.Builder.configuration()
                    .withAndroidSdk(parsedArgs.sdk ?: Utils.cleanFileSafe(CommonDefaults.ANDROID_SDK))
                    .withApplicationApk(parsedArgs.apk)
                    .withApplicationPackage(tongsConfiguration.applicationPackage)
                    .withInstrumentationApk(parsedArgs.testApk)
                    .withInstrumentationPackage(tongsConfiguration.instrumentationPackage)
                    .withOutput(Utils.cleanFileSafe(tongsConfiguration.baseOutputDir ?: CommonDefaults.DEFAULT_OUTPUT))
                    .withTitle(tongsConfiguration.title)
                    .withSubtitle(tongsConfiguration.subtitle)
                    .withTestPackage(tongsConfiguration.testPackage)
                    .withPlugins(tongsConfiguration.plugins)
                    .withTestOutputTimeout(tongsConfiguration.testOutputTimeout)
                    .withTestRunnerClass(tongsConfiguration.android.testRunnerClass)
                    .withTestRunnerArguments(tongsConfiguration.android.instrumentationArguments)
                    .withExcludedSerials(tongsConfiguration.excludedSerials)
                    .withFallbackToScreenshots(tongsConfiguration.fallbackToScreenshots)
                    .withTotalAllowedRetryQuota(tongsConfiguration.totalAllowedRetryQuota)
                    .withRetryPerTestCaseQuota(tongsConfiguration.retryPerTestCaseQuota)
                    .withCoverageEnabled(tongsConfiguration.isCoverageEnabled)
                    .withPoolingStrategy(tongsConfiguration.poolingStrategy)
                    .withExcludedAnnotation(tongsConfiguration.excludedAnnotation)
                    .withTongsIntegrationTestRunType(tongsConfiguration.tongsIntegrationTestRunType)
                    .withPluginConfiguration(tongsConfiguration.configuration)
                    .build(true)

            val tongs = Tongs(configuration)
            if (!tongs.run() && !tongsConfiguration.ignoreFailures) {
                exitProcess(1)
            }
        } catch (e: FileNotFoundException) {
            logger.error("Could not find configuration file", e)
            exitProcess(1)
        }
    }

    class CommandLineArgs {
        @JvmField
        @field:Parameter(names = ["--sdk"], description = "Path to Android SDK", converter = FileConverter::class)
        var sdk: File? = null

        @field:Parameter(names = ["--apk"], description = "Application APK", converter = FileConverter::class)
        var apk: File? = null

        @field:Parameter(names = ["--test-apk"], description = "Test application APK", converter = FileConverter::class)
        var testApk: File? = null

        @field:Parameter(names = ["--config"], description = "Path of JSON config file", required = true)
        lateinit var configurationFile: String

        @JvmField
        @field:Parameter(names = ["-h", "--help"], description = "Command help", help = true, hidden = true)
        var help = false
    }

    /* JCommander deems it necessary that this class be public. Lame. */
    class FileConverter : IStringConverter<File?> {
        override fun convert(s: String): File? {
            return Utils.cleanFile(s)
        }
    }
}